### PR TITLE
IWAID - 8 BE Crud prescripciones

### PR DESCRIPTION
**Pull request by IrvingF-4**
**Ticket:**

- [IWAID-8](https://iwaids.atlassian.net/browse/IWAID-8)

**Overview:**
The endpoint for managing medical prescriptions has been implemented. And also, the file for creating the table of medical prescriptions has been added.

**What a change:**
The endpoint for handling medical histories has been developed, which allows retrieving all of them, a specific one by ID, updating, and deleting.

**How to test:**
To test the endpoint, it is necessary to run the project and open Postman.

- To test by fetching all the registered prescriptions, you need to create a new GET request with the following URL: http://localhost:8080/iwaid/prescription, and send the request, which will return a JSON with all the created prescriptions and a status of 200.
- To retrieve a specific prescription, you need to create a GET request with the following URL: [http://localhost:8080/iwaid/prescription/{id}](http://localhost:8080/iwaid/prescription/%7Bid%7D), where id is replaced with the desired prescription's ID. When you send the request, it will return all the data of the prescription, if it's found, with a status of 200.
- To save a new prescription, you need to create a new POST request with the following URL: http://localhost:8080/iwaid/prescription and include a JSON in the request body like the following:                                                          
{
    "patientId": 1,
    "doctorId": 1,
    "registerDate": "2023-08-05",
    "description": ""
}                                                                                                                                                                                                   
 If there are no issues with the data saved when you send the request, it will return a JSON with the prescription data and a status of 201.

- To update a prescription, you need to create a new PATCH request with the following URL: [http://localhost:8080/iwaid/prescription/{id}](http://localhost:8080/iwaid/prescription/%7Bid%7D), where {id} is the ID of the prescription you want to update. Include a JSON in the request body as follows:                                                                           
 {
    "patientId": 3,
    "doctorId": 1,
    "registerDate": "2023-08-05",
    "description": "Data to be changed"
}                                                                                                                                                                                                         
 By sending the request, it will update the prescription with the provided data, and if there are no issues, it will return a JSON with the updated prescription data and a status of 200.

- To delete a prescription, you need to create a new DELETE request with the following URL: [http://localhost:8080/iwaid/prescription/{id}](http://localhost:8080/iwaid/prescription/%7Bid%7D), where {id} is the ID of the prescription you want to delete. If everything goes well, it will send a status of 200.

**Evidence:**
- Get all prescriptions:
 
 ![image](https://github.com/DavidVelaz17/iwaid-api/assets/85428204/0ceae266-bdd1-4e82-8efb-80bbac3a078e)

- Search Prescription by ID:
 
 ![image](https://github.com/DavidVelaz17/iwaid-api/assets/85428204/c1b48a3b-3340-4656-998a-60c460474201)

- Create a new prescription:
 
![image](https://github.com/DavidVelaz17/iwaid-api/assets/85428204/d8e09d88-31fe-4164-8fab-061a36ae8f09)

- Update a prescription:
 
![image](https://github.com/DavidVelaz17/iwaid-api/assets/85428204/4d5ac4d2-e96f-4f6f-bdb1-f4be716e1f06)

- Delete a prescription

![image](https://github.com/DavidVelaz17/iwaid-api/assets/85428204/0e99fe0f-9ed4-4194-964e-ce690fe5d7ae)

- Creation of the prescription file for flyway:
![image](https://github.com/DavidVelaz17/iwaid-api/assets/85428204/cef7c146-512b-4ee5-9a57-0cc3f6559cc3)

### CODE REVIEW

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/dto/PrescriptionDTO.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    private PatientDTO patient;
+
+    @ApiObjectField(name = "doctorId", description = "Doctor's ID")
+    private int doctorId;
+
+    @ApiObjectField(name = "doctor", description = "Doctor's data")
+    private DoctorDTO doctor;
+
+    @ApiObjectField(name = "date", description = "Prescription date")
+    private LocalDate date;
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/dto/PrescriptionDTO.java]: esta fecha indica la fecha de registro? entonces su nombre correcto sería registerDate

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/dto/PrescriptionDTO.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    @ApiObjectField(name = "patient", description = "Patient's Data")
+    private PatientDTO patient;
+
+    @ApiObjectField(name = "doctorId", description = "Doctor's ID")
+    private int doctorId;
+
+    @ApiObjectField(name = "doctor", description = "Doctor's data")
+    private DoctorDTO doctor;
+
+    @ApiObjectField(name = "date", description = "Prescription date")
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/dto/PrescriptionDTO.java]: actualiza descripción

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/dto/PrescriptionDTO.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    private LocalDate date;
+
+    @ApiObjectField(name = "description", description = "Prescription description")
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/dto/PrescriptionDTO.java]:
```suggestion
    @ApiObjectField(name = "description", description = "Prescription's description")
```

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/PrescriptionController.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    public PrescriptionController(PrescriptionService service) {
+        this.service = service;
+    }
+
+    @GetMapping
+    public ResponseEntity<List<PrescriptionDTO>> getAllPrescription() throws Exception {
+        return new ResponseEntity<>(service.getAllPrescriptions(), HttpStatus.OK);
+    }
+
+    @GetMapping("/{id}")
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/PrescriptionController.java]: id de que? renombrar

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/PrescriptionController.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    @PostMapping
+    public ResponseEntity<PrescriptionDTO> createPrescription(@Valid @RequestBody final PrescriptionForm form)
+            throws Exception {
+        PrescriptionDTO prescription = service.createPrescription(form);
+        return new ResponseEntity<>(prescription, HttpStatus.CREATED);
+    }
+
+    @PatchMapping("/{id}")
+    public ResponseEntity<PrescriptionDTO> updatePrescription(@Valid @RequestBody final PrescriptionForm form,
+                                                              @PathVariable final int id) throws Exception {
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/PrescriptionController.java]: ditto

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/PrescriptionController.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        PrescriptionDTO prescription = service.updatePrescription(form, id);
+        return new ResponseEntity<>(prescription, HttpStatus.OK);
+    }
+
+    @DeleteMapping("/{id}")
+    public ResponseEntity<Void> deletePrescription(@PathVariable final int id) throws Exception {
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/PrescriptionController.java]: ditto

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/PrescriptionService.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+                        .build(prescription,
+                                patientDTOMap
+                                        .get(prescription.getPatientId()),
+                                doctorDTOMap
+                                        .get(prescription.getDoctorId())
+                        ))
+                .toList();
+    }
+
+    public PrescriptionDTO findPrescriptionById(final int id) throws Exception {
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/PrescriptionService.java:43]: prescriptionId

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/PrescriptionService.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+    public PrescriptionDTO createPrescription(final PrescriptionForm form) throws Exception {
+        final Prescription prescription = new Prescription(form);
+        patientService.validateIfPatientExist(form.getPatientId());
+        doctorService.validateIfDoctorExists(form.getDoctorId());
+        repository.save(prescription);
+        return PrescriptionDTO.build(prescription);
+    }
+
+    public PrescriptionDTO updatePrescription(final PrescriptionForm form, final int id)
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/PrescriptionService.java]: prescriptionId

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/PrescriptionService.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        validateIfPrescriptionExist(id);
+        final Prescription prescription = repository.findById(id).orElseThrow();
+        patientService.validateIfPatientExist(form.getPatientId());
+        doctorService.validateIfDoctorExists(form.getDoctorId());
+        prescription.updatePrescription(form);
+        repository.save(prescription);
+        return PrescriptionDTO.build(prescription);
+    }
+
+    public void deletePrescription(final int id) throws Exception {
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/PrescriptionService.java]: prescriptionId

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/PrescriptionService.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        validateIfPrescriptionExist(id);
+        repository.deleteById(id);
+    }
+
+    public void validateIfPrescriptionExist(final int id) throws Exception {
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/PrescriptionService.java]: prescriptionId

File: consultorio-medico/src/main/resources/db/migration/V1_005__Prescriptions.sql [hunk 11]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE IF NOT EXISTS `prescriptions` (
+    `id` INT NOT NULL AUTO_INCREMENT,
+    `patients_id` INT NOT NULL,
```
Comment (SendyCruz) [inline consultorio-medico/src/main/resources/db/migration/V1_005__Prescriptions.sql]:
```suggestion
    `patient_id` INT NOT NULL,
```

File: consultorio-medico/src/main/resources/db/migration/V1_005__Prescriptions.sql [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+    `doctors_id` INT NOT NULL,
```
Comment (SendyCruz) [inline consultorio-medico/src/main/resources/db/migration/V1_005__Prescriptions.sql]:
```suggestion
    `doctor_id` INT NOT NULL,
```

**Review by SendyCruz — APPROVED** [review]

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/DoctorService.java [hunk 13]
```diff
@@ -46,9 +49,19 @@ public DoctorDTO updateDoctor(final DoctorForm form, final int id) throws Except
         return DoctorDTO.build(doctor);
     }
 
-    private void validateIfDoctorExists(int id) throws Exception {
+    public Map<Integer, DoctorDTO> getDoctorsByIds(final List<Integer> doctorIds) {
+        final List<Doctor> doctors = doctorRepository.findAllById(doctorIds);
+        return doctorDTOs(doctors);
+    }
+
+    public void validateIfDoctorExists(int id) throws Exception {
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/DoctorService.java]: doctorId

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/PatientService.java [hunk 14]
```diff
@@ -53,9 +56,19 @@ public void deletePatient(final int id) throws Exception {
         patientRepository.deleteById(id);
     }
 
+    public Map<Integer, PatientDTO> getPatientByIds(final List<Integer> patientsId) {
+        final List<Patient> patients = patientRepository.findAllById(patientsId);
+        return patientDTOs(patients);
+    }
+
     public void validateIfPatientExist(final int id) throws Exception {
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/PatientService.java]: patientId

**Review by SendyCruz — CHANGES_REQUESTED** [review]

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/PrescriptionController.java [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+import jakarta.validation.Valid;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping(path = "/iwaid/prescription")
```
Comment (cgarcia-iwa) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/PrescriptionController.java]: definan un estandar de las rutas del back, unas estan terminando con / y otras no

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/PrescriptionService.java [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+    public PrescriptionDTO findPrescriptionById(final int id) throws Exception {
+        validateIfPrescriptionExist(id);
+        final Prescription prescription = repository.findById(id).orElseThrow();
+        final PatientDTO patientDTO = patientService.getPatientById(prescription.getPatientId());
+        final DoctorDTO doctorDTO = doctorService.getDoctorById(prescription.getDoctorId());
+        return PrescriptionDTO.build(prescription, patientDTO, doctorDTO);
+    }
+
+    public PrescriptionDTO createPrescription(final PrescriptionForm form) throws Exception {
+        final Prescription prescription = new Prescription(form);
```
Comment (cgarcia-iwa) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/PrescriptionService.java]: moverlo despues de las validaciones, replicar en casos similares

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/PrescriptionService.java [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    public void deletePrescription(final int prescriptionId) throws Exception {
+        validateIfPrescriptionExist(prescriptionId);
+        repository.deleteById(prescriptionId);
+    }
+
+    public void validateIfPrescriptionExist(final int prescriptionId) throws Exception {
+        if (!repository.existsById(prescriptionId)) {
+            throw new Exception("Prescription not found with ID: " + prescriptionId);
```
Comment (cgarcia-iwa) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/PrescriptionService.java]: determinar estandar para los errores

**Review by cgarcia-iwa — CHANGES_REQUESTED** [review]

**Comment by SendyCruz** [comment]
resolver conflictos

**Review by SendyCruz — APPROVED** [review]

**Review by cgarcia-iwa — APPROVED** [review]

**Pull request merged**